src/main/java/com/learning/mfscreener/service/CapitalGainsService.java

```diff
@@ -0,0 +1,134 @@
+package com.learning.mfscreener.service;
+
+import com.learning.mfscreener.config.logging.Loggable;
+import com.learning.mfscreener.exception.GainsException;
+import com.learning.mfscreener.exception.IncompleteCASError;
+import com.learning.mfscreener.models.portfolio.CasDTO;
+import com.learning.mfscreener.models.portfolio.Fund;
+import com.learning.mfscreener.models.portfolio.FundType;
+import com.learning.mfscreener.models.portfolio.GainEntry;
+import com.learning.mfscreener.models.portfolio.UserFolioDTO;
+import com.learning.mfscreener.models.portfolio.UserSchemeDTO;
+import com.learning.mfscreener.models.portfolio.UserTransactionDTO;
+import com.learning.mfscreener.utils.FIFOUnits;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CapitalGainsService {
+
+    private static final double MIN_OPEN_BALANCE = 0.01;
+
+    private BigDecimal investedAmount = BigDecimal.ZERO;
+    private Double currentValue = 0D;
+    private final List<GainEntry> gainEntries = new ArrayList<>();
+    List<String> errors = new ArrayList<>();
+
+    @Loggable(params = false)
+    Map<String, Object> processData(CasDTO casDTO) throws IncompleteCASError {
+
+        for (UserFolioDTO folio : casDTO.folios()) {
+            for (UserSchemeDTO scheme : folio.schemes()) {
+                List<UserTransactionDTO> transactions = scheme.transactions();
+                Fund fund = new Fund(scheme.scheme(), folio.folio(), scheme.isin(), scheme.type());
+                if (!transactions.isEmpty()) {
+                    if (Double.parseDouble(scheme.myopen()) >= MIN_OPEN_BALANCE) {
+                        throw new IncompleteCASError(
+                                "Incomplete CAS found. For gains computation all folios should have zero opening balance");
+                    }
+                    try {
+                        FIFOUnits fifo = new FIFOUnits(fund, transactions);
+                        investedAmount = this.investedAmount.add(fifo.getInvested());
+                        currentValue = currentValue + scheme.valuation().value();
+                        gainEntries.addAll(fifo.getGains());
+                    } catch (GainsException exc) {
+                        this.errors.add(fund.scheme() + ", " + exc.getMessage());
+                    }
+                }
+            }
+        }
+        return prepareGains(gainEntries);
+    }
+
+    Map<String, Object> prepareGains(List<GainEntry> gainEntries) {
```

Comment:
@coderabbitai, refactor this method to a more readable format without loss of functionality.

Reply:
@coderabbitai, I would like to refactor prepareGains method as well